File: sdk-test-go/tests/cloudwatch_test.go
package tests

import (
	"context"
	"testing"
	"time"

	"floci-sdk-test-go/internal/testutil"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/cloudwatch"
	cwtypes "github.com/aws/aws-sdk-go-v2/service/cloudwatch/types"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestCloudWatch(t *testing.T) {
	ctx := context.Background()
	svc := testutil.CloudWatchClient()
	namespace := "GoTest"

	t.Run("PutMetricData", func(t *testing.T) {
		_, err := svc.PutMetricData(ctx, &cloudwatch.PutMetricDataInput{
			Namespace: aws.String(namespace),
			MetricData: []cwtypes.MetricDatum{
				{
					MetricName: aws.String("RequestCount"),
					Value:      aws.Float64(42),
					Unit:       cwtypes.StandardUnitCount,
					Timestamp:  aws.Time(time.Now()),
				},
			},
		})
		require.NoError(t, err)
	})

	t.Run("ListMetrics", func(t *testing.T) {
		r, err := svc.ListMetrics(ctx, &cloudwatch.ListMetricsInput{Namespace: aws.String(namespace)})
		require.NoError(t, err)
		assert.NotEmpty(t, r.Metrics)
	})

	t.Run("GetMetricStatistics", func(t *testing.T) {
		now := time.Now()
		r, err := svc.GetMetricStatistics(ctx, &cloudwatch.GetMetricStatisticsInput{
			Namespace:  aws.String(namespace),
			MetricName: aws.String("RequestCount"),
			StartTime:  aws.Time(now.Add(-5 * time.Minute)),
			EndTime:    aws.Time(now.Add(time.Minute)),
			Period:     aws.Int32(60),
			Statistics: []cwtypes.Statistic{cwtypes.StatisticSum},
		})
		require.NoError(t, err)
		assert.NotEmpty(t, r.Datapoints)
	})
}

File: sdk-test-rust/tests/cloudwatch_test.rs
mod common;

use aws_sdk_cloudwatch::types::{MetricDatum, StandardUnit, Statistic};

#[tokio::test]
async fn test_cloudwatch_put_metric_data() {
    let cw = common::cloudwatch_client().await;
    let namespace = "RustTest";

    let result = cw
        .put_metric_data()
        .namespace(namespace)
        .metric_data(
            MetricDatum::builder()
                .metric_name("RequestCount")
                .value(42.0)
                .unit(StandardUnit::Count)
                .build(),
        )
        .send()
        .await;
    assert!(result.is_ok(), "PutMetricData failed: {:?}", result.err());
}

#[tokio::test]
async fn test_cloudwatch_list_metrics() {
    let cw = common::cloudwatch_client().await;
    let namespace = "RustTestList";

    // Setup
    cw.put_metric_data()
        .namespace(namespace)
        .metric_data(
            MetricDatum::builder()
                .metric_name("TestMetric")
                .value(1.0)
                .unit(StandardUnit::Count)
                .build(),
        )
        .send()
        .await
        .expect("setup");

    let result = cw.list_metrics().namespace(namespace).send().await;
    assert!(result.is_ok(), "ListMetrics failed: {:?}", result.err());
    assert!(!result.unwrap().metrics().is_empty());
}

#[tokio::test]
async fn test_cloudwatch_get_metric_statistics() {
    let cw = common::cloudwatch_client().await;
    let namespace = "RustTestStats";

    // Setup
    cw.put_metric_data()
        .namespace(namespace)
        .metric_data(
            MetricDatum::builder()
                .metric_name("StatsMetric")
                .value(100.0)
                .unit(StandardUnit::Count)
                .build(),
        )
        .send()
        .await
        .expect("setup");

    let now = std::time::SystemTime::now();
    let five_mins_ago = now - std::time::Duration::from_secs(300);
    let one_min_future = now + std::time::Duration::from_secs(60);

    let result = cw
        .get_metric_statistics()
        .namespace(namespace)
        .metric_name("StatsMetric")
        .start_time(aws_smithy_types::DateTime::from(five_mins_ago))
        .end_time(aws_smithy_types::DateTime::from(one_min_future))
        .period(60)
        .statistics(Statistic::Sum)
        .send()
        .await;
    assert!(result.is_ok(), "GetMetricStatistics failed: {:?}", result.err());
    assert!(!result.unwrap().datapoints().is_empty());
}
